File: packages/plugin-bootstrap/python/elizaos_plugin_bootstrap/actions/send_message.py
"""
SEND_MESSAGE Action - Send a message to a specific target.

This action allows the agent to send messages to specific
rooms, entities, or channels.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING
from uuid import UUID

from elizaos.types import Action, ActionExample, ActionResult, Content, MemoryType

if TYPE_CHECKING:
    from elizaos.types import HandlerCallback, HandlerOptions, IAgentRuntime, Memory, State


@dataclass
class SendMessageAction:
    """
    Action for sending messages to specific targets.

    This action is used when:
    - A message needs to be sent to a specific room
    - Direct messaging to an entity is required
    - Cross-room communication is needed
    """

    name: str = "SEND_MESSAGE"
    similes: list[str] = field(
        default_factory=lambda: [
            "MESSAGE",
            "DM",
            "DIRECT_MESSAGE",
            "POST_MESSAGE",
            "NOTIFY",
        ]
    )
    description: str = (
        "Send a message to a specific room or entity. "
        "Use this for targeted communication outside the current context."
    )

    async def validate(self, runtime: IAgentRuntime, message: Memory) -> bool:
        """Validate that message can be sent."""
        # Check if target information is available
        if message.content and message.content.target:
            return True
        return True  # Default validation passes

    async def handler(
        self,
        runtime: IAgentRuntime,
        message: Memory,
        state: State | None = None,
        options: HandlerOptions | None = None,
        callback: HandlerCallback | None = None,
        responses: list[Memory] | None = None,
    ) -> ActionResult:
        """Handle sending a message."""
        # Get message content from responses
        message_text = ""
        if responses and responses[0].content:
            message_text = str(responses[0].content.text or "")

        if not message_text:
            return ActionResult(
                text="No message content to send",
                values={"success": False, "error": "no_content"},
                data={"actionName": "SEND_MESSAGE"},
                success=False,
            )

        # Determine target
        target_room_id = message.room_id
        target_entity_id: UUID | None = None

        if message.content and message.content.target:
            target = message.content.target
            if isinstance(target, dict):
                room_str = target.get("roomId")
                entity_str = target.get("entityId")
                if room_str:
                    try:
                        target_room_id = UUID(room_str)
                    except ValueError:
                        pass
                if entity_str:
                    try:
                        target_entity_id = UUID(entity_str)
                    except ValueError:
                        pass

        if not target_room_id:
            return ActionResult(
                text="No target room specified",
                values={"success": False, "error": "no_target"},
                data={"actionName": "SEND_MESSAGE"},
                success=False,
            )

        try:
            # Create the message memory
            message_content = Content(
                text=message_text,
                source="agent",
                actions=["SEND_MESSAGE"],
            )

            await runtime.create_memory(
                content=message_content,
                room_id=target_room_id,
                entity_id=runtime.agent_id,
                memory_type=MemoryType.MESSAGE,
                metadata={
                    "type": "SEND_MESSAGE",
                    "targetEntityId": str(target_entity_id) if target_entity_id else None,
                },
            )

            response_content = Content(
                text=f"Message sent: {message_text[:50]}...",
                actions=["SEND_MESSAGE"],
            )

            if callback:
                await callback(response_content)

            return ActionResult(
                text=f"Message sent to room",
                values={
                    "success": True,
                    "messageSent": True,
                    "targetRoomId": str(target_room_id),
                    "targetEntityId": str(target_entity_id) if target_entity_id else None,
                },
                data={
                    "actionName": "SEND_MESSAGE",
                    "targetRoomId": str(target_room_id),
                    "messagePreview": message_text[:100],
                },
                success=True,
            )

        except Exception as error:
            runtime.logger.error(
                {
                    "src": "plugin:bootstrap:action:sendMessage",
                    "agentId": runtime.agent_id,
                    "error": str(error),
                },
                "Error sending message",
            )
            return ActionResult(
                text="Error sending message",
                values={"success": False, "error": str(error)},
                data={"actionName": "SEND_MESSAGE", "error": str(error)},
                success=False,
                error=error,
            )

    @property
    def examples(self) -> list[list[ActionExample]]:
        """Example interactions demonstrating the SEND_MESSAGE action."""
        return [
            [
                ActionExample(
                    name="{{name1}}",
                    content=Content(text="Send a hello to the general channel."),
                ),
                ActionExample(
                    name="{{name2}}",
                    content=Content(
                        text="I'll send that message now.",
                        actions=["SEND_MESSAGE"],
                    ),
                ),
            ],
        ]


# Create the action instance
send_message_action = Action(
    name=SendMessageAction.name,
    similes=SendMessageAction().similes,
    description=SendMessageAction.description,
    validate=SendMessageAction().validate,
    handler=SendMessageAction().handler,
    examples=SendMessageAction().examples,
)

File: plugins/plugin-bootstrap/rust/src/actions/send_message.rs
//! SEND_MESSAGE action implementation.

use async_trait::async_trait;
use std::collections::HashMap;
use std::sync::Arc;
use uuid::Uuid;

use crate::error::{PluginError, PluginResult};
use crate::runtime::IAgentRuntime;
use crate::types::{ActionResult, Content, Memory, MemoryType, State};

use super::Action;

/// Action for sending messages.
pub struct SendMessageAction;

#[async_trait]
impl Action for SendMessageAction {
    fn name(&self) -> &'static str {
        "SEND_MESSAGE"
    }

    fn similes(&self) -> &[&'static str] {
        &["MESSAGE", "DM", "DIRECT_MESSAGE", "POST_MESSAGE", "NOTIFY"]
    }

    fn description(&self) -> &'static str {
        "Send a message to a specific room or entity. \
         Use this for targeted communication outside the current context."
    }

    async fn validate(&self, _runtime: &dyn IAgentRuntime, _message: &Memory) -> bool {
        true
    }

    async fn handler(
        &self,
        runtime: Arc<dyn IAgentRuntime>,
        message: &Memory,
        _state: Option<&State>,
        responses: Option<&[Memory]>,
    ) -> PluginResult<ActionResult> {
        // Get message content from responses
        let message_text = responses
            .and_then(|r| r.first())
            .and_then(|r| Some(r.content.text.clone()))
            .filter(|t| !t.is_empty())
            .ok_or_else(|| PluginError::InvalidInput("No message content to send".to_string()))?;

        // Determine target
        let target_room_id = message
            .content
            .target
            .as_ref()
            .and_then(|t| t.room_id)
            .or(message.room_id)
            .ok_or_else(|| PluginError::InvalidInput("No target room specified".to_string()))?;

        let target_entity_id: Option<Uuid> = message
            .content
            .target
            .as_ref()
            .and_then(|t| t.entity_id);

        // Create the message memory
        let mut metadata = HashMap::new();
        metadata.insert("type".to_string(), serde_json::json!("SEND_MESSAGE"));
        if let Some(entity_id) = target_entity_id {
            metadata.insert("targetEntityId".to_string(), serde_json::json!(entity_id.to_string()));
        }

        runtime
            .create_memory(
                Content {
                    text: message_text.clone(),
                    actions: vec!["SEND_MESSAGE".to_string()],
                    ..Default::default()
                },
                Some(target_room_id),
                Some(runtime.agent_id()),
                MemoryType::Message,
                metadata,
            )
            .await?;

        let preview = if message_text.len() > 50 {
            format!("{}...", &message_text[..50])
        } else {
            message_text.clone()
        };

        Ok(ActionResult::success("Message sent")
            .with_value("success", true)
            .with_value("messageSent", true)
            .with_value("targetRoomId", target_room_id.to_string())
            .with_data("actionName", "SEND_MESSAGE")
            .with_data("targetRoomId", target_room_id.to_string())
            .with_data("messagePreview", preview))
    }
}

